domain/application/service/application.go

```diff
@@ -335,6 +347,37 @@ func validateCreateApplicationParams(
 	if err := origin.Validate(); err != nil {
 		return fmt.Errorf("%w: %v", applicationerrors.CharmOriginNotValid, err)
 	}
+
+	// Validate consistency of resources origin and revision
+	if err := resolvedResources.Validate(); err != nil {
+		return err
+	}
+
+	// Validates that all charm resources are resolved
+	appResourceSet := set.NewStrings()
+	charmResourceSet := set.NewStrings()
+	for _, res := range charm.Meta().Resources {
+		charmResourceSet.Add(res.Name)
+	}
+	for _, res := range resolvedResources {
+		appResourceSet.Add(res.Name)
+	}
+	unexpectedResources := appResourceSet.Difference(charmResourceSet)
+	missingResources := charmResourceSet.Difference(appResourceSet)
```

Comment:
This is an FYI more than anything, these are left-biased difference methods, in other words, it will tell you what on the left-hand side is different, but not both (i.e. the right-hand side). It's not comprehensive.